lib/enum.js

```diff
@@ -2,52 +2,96 @@
 
 class Enum {}
 
-const enumArray = values => class extends Enum {
-  constructor(val) {
-    super();
-    this.value = this.constructor.key(val);
+const enumArray = values => {
+  const keysIndex = new Map();
+  const enumValues = new Map();
+  for (let i = 0; i < values.length; ++i) {
+    keysIndex.set(values[i], i);
   }
-  static get collection() {
-    return values;
-  }
-  static has(val) {
-    return values.includes(val);
-  }
-  static key(val) {
-    return values.includes(val) ? val : undefined;
-  }
-  [Symbol.toPrimitive]() {
-    return this.value;
+  const enumClass = class extends Enum {
+    constructor(val) {
+      super();
+      this.keyValue = val;
+      this.value = this.constructor.key(val);
+      if (this.value === undefined) {
+        throw new TypeError(
```

Comment:
I think that this will make use of enums much easier to debug as it will point to the issue exactly instead of having to search why some check (for example enum comparison) silently failed because you've made a typo when creating an enum. Furthermore, this shouldn't be triggered usually (only during debug/testing) as it's not normal to create enum from non-existent enum-values.

Also it is incorrect to have such enum that 
> so value can be just undefined if somebody pass unexpected value to constructor.

because that's basically another value for enum, but it must not have any other values besides those provided at construction time.

Reply:
Does `Number('hello')` throws?

Reply:
`Enum.neither` or propose other synonym

Reply:
What about Enum.NaE (not an enum)?